pkg/pillar/cmd/volumemgr/updatestatus.go

```diff
@@ -483,15 +484,55 @@ func doUpdateVol(ctx *volumemgrContext, status *types.VolumeStatus) (bool, bool)
 			status.ReferenceName = ctStatus.ReferenceID()
 			status.ContentFormat = ctStatus.Format
 			changed = true
-			// Asynch creation; ensure we have requested it
-			AddWorkCreate(ctx, status)
+			// Asynch preparation; ensure we have requested it
+			AddWorkPrepare(ctx, status)
 		}
 		if status.IsErrorSource(types.ContentTreeStatus{}) {
 			log.Functionf("doUpdate: Clearing volume error %s", status.Error)
 			status.ClearErrorWithSource()
 			changed = true
 		}
-		if status.State == types.CREATING_VOLUME && !status.VolumeCreated {
+		if status.State == types.CREATING_VOLUME && !status.VolumeCreated && !status.VolumePreparing && !status.VolumePrepareDone {
+			vr := popVolumePrepareResult(ctx, status.Key())
+			if vr != nil {
+				log.Functionf("doUpdateVol: VolumePrepareResult(%s)", status.Key())
+				if vr.Error != nil {
+					log.Errorf("doUpdateVol: Error received from the volume prepare worker %v",
+						vr.Error)
+					status.SetErrorWithSource(vr.Error.Error(),
+						types.VolumeStatus{}, vr.ErrorTime)
+					changed = true
+					return changed, false
+				} else if status.IsErrorSource(types.VolumeStatus{}) {
+					log.Functionf("doUpdateVol: Clearing volume error %s", status.Error)
+					status.ClearErrorWithSource()
+					changed = true
+				}
+				status.VolumePreparing = true
+				changed = true
+			} else {
+				log.Functionf("doUpdateVol: VolumePrepareResult(%s) not found", status.Key())
+			}
+		}
+		if status.State == types.CREATING_VOLUME && !status.VolumeCreated && status.VolumePreparing && !status.VolumePrepareDone {
+			if ctx.persistType == types.PersistZFS && !status.IsContainer() {
+				for _, s := range ctx.subZVolStatus.GetAll() {
+					zVolStatus := s.(types.ZVolStatus)
+					if zVolStatus.Dataset == status.ZVolName(types.VolumeZFSPool) {
```

Comment:
Makes sense to define this as a lookup function if you don't already have one.

Reply:
Done